crates/gql/src/query.rs

```diff
@@ -26,53 +32,156 @@ impl GraphQLType<DefaultScalarValue> for Query {
     where
         DefaultScalarValue: 'r,
     {
+        let _ = registry.get_type::<&BigInt>(&());
+
         let mut fields = vec![];
         fields.push(registry.field::<&String>("apiVersion", &()));
 
-        for (name, entry) in spec.entries.iter() {
-            fields.push(registry.field::<&[DynamicSchemaMetadata]>(name, &entry));
+        for (name, schema_spec) in spec.entries.iter() {
+            let filter = registry.arg::<Option<SubgraphFilterSpec>>("filter", &schema_spec.filter);
+
+            let field = registry
+                .field::<&[DynamicSchemaSpec]>(name, &schema_spec)
+                .argument(filter);
+            fields.push(field);
         }
+
         registry
-            .build_object_type::<Query>(&spec, &fields)
+            .build_object_type::<DynamicQuery>(&spec, &fields)
             .into_meta()
     }
 }
 
-impl GraphQLValue<DefaultScalarValue> for Query {
-    type Context = Context;
+#[derive(Debug, Clone)]
+pub struct MemoryStore {
+    /// A map of subgraph UUIDs to their names
+    pub subgraph_name_lookup: Arc<RwLock<BTreeMap<Uuid, String>>>,
+    /// A map of subgraph names to their entries
+    pub entries_store: Arc<RwLock<BTreeMap<String, (Uuid, Vec<SubgraphSpec>)>>>,
+    // A broadcaster for entry updates
+    // pub entries_broadcaster: tokio::sync::broadcast::Sender<SubgraphDataEntryUpdate>,
+}
+
+impl MemoryStore {
+    pub fn new() -> MemoryStore {
+        MemoryStore {
+            subgraph_name_lookup: Arc::new(RwLock::new(BTreeMap::new())),
+            entries_store: Arc::new(RwLock::new(BTreeMap::new())),
+            // entries_broadcaster,
+        }
+    }
+}
+
+impl Dataloader for MemoryStore {
+    fn fetch_entries_from_subgraph(
+        &self,
+        subgraph_name: &str,
+        _executor: &Executor<DataloaderContext>,
+        _schema: &DynamicSchemaSpec,
+    ) -> Result<Vec<SubgraphSpec>, FieldError> {
+        let subgraph_db = self
+            .entries_store
+            .read()
+            .expect("failed to read from gql entries store");
```

Comment:
It's hard to tell from my context if this is a safe `expect`. I want to avoid the surfpool Terminal UI being mucked up with weird errors because a thread is killed. I'll leave it to your judgement if this is safe or not.

Reply:
Fixed, thanks!